### PR TITLE
refactor: add LedgerFacade with new entry builder

### PR DESCRIPTION
**Pull request by UncleSamtoshi**
Continues the work of PR #1170:
- add ledger facade
- rewrite ledger entry-builder to work with expected PaymentFlow updates

### CODE REVIEW

File: src/domain/ledger/index.types.d.ts [hunk 1]
```diff
@@ -124,22 +124,22 @@ type AddColdStorageTxSendArgs = {
...
-  senderUsername: Username | null
+  senderUsername?: Username
   description: string
   sats?: Satoshis
   cents?: UsdCents
   recipientWalletId: WalletId
   recipientWalletCurrency: WalletCurrency
-  recipientUsername: Username | null
-  memoPayer: string | null
+  recipientUsername?: Username
```
Comment (UncleSamtoshi) [inline src/domain/ledger/index.types.d.ts:133]: @dolcalmi Any thoughts on changing these two metadata value types from null to undefined?

Comment (UncleSamtoshi) [reply src/domain/ledger/index.types.d.ts:133]: This is the place where transactions are brought from mongoose into our domain https://github.com/GaloyMoney/galoy/blob/21b76e1e261ea818f03d42a182b044a50397da8c/src/services/ledger/index.ts#L308

Comment (UncleSamtoshi) [reply src/domain/ledger/index.types.d.ts:133]: The current [LedgerTransaction type](https://github.com/GaloyMoney/galoy/blob/e9a9686ee704bd6e7cc09834d09c266be9c58739/src/domain/ledger/index.types.d.ts#L24) actually already uses undefined, but I'm not sure how a null value persisted in the database becomes an undefined value.

**Review by bodymindarts — DISMISSED** [review]

File: src/app/wallets/send-intraledger.ts [hunk 2]
```diff
@@ -230,7 +230,7 @@ const executePaymentViaIntraledger = async ({
           amountDisplayCurrency,
           recipientWalletId,
           recipientWalletCurrency: recipientWallet.currency,
-          recipientUsername,
+          recipientUsername: recipientUsername ?? undefined,
```
Comment (dolcalmi) [inline src/app/wallets/send-intraledger.ts:233]: are you sure we want to exclude `""`? else use `||`

File: src/app/wallets/send-on-chain.ts [hunk 3]
```diff
@@ -241,7 +241,7 @@ const executePaymentViaIntraledger = async ({
             recipientWalletId: recipientWallet.id,
             recipientWalletCurrency: recipientWallet.currency,
             recipientUsername: recipientAccount.username,
-            memoPayer: memo ?? null,
+            memoPayer: memo ?? undefined,
```
Comment (dolcalmi) [inline src/app/wallets/send-on-chain.ts:244]: are you sure we want to exclude `""`? else please use `||`

Comment (UncleSamtoshi) [reply src/app/wallets/send-on-chain.ts:244]: I think it makes sense to exclude ```""```. Can you think of any scenario where we would want it?

File: src/services/ledger/domain/entry-builder.ts [hunk 4]
```diff
@@ -1,194 +1,195 @@
...
         entry,
         metadata,
         staticAccountIds,
-        btcAmount: amount,
-        usdAmount: debitAmount,
-      })
+        btcAmount: calc.sub(btcWithFee, btcWithOutFee),
+        usdAmount: calc.sub(usdWithFee, usdWithOutFee),
+      }
+      withdrawUsdFromDealer(dealerData)
```
Comment (dolcalmi) [inline src/services/ledger/domain/entry-builder.ts:158]: this is a bad practice (it is really difficult to follow what is the purpose of the method) or we need to change the method name or we need to update entry in this line.

File: src/services/ledger/domain/entry-builder.ts [hunk 5]
```diff
@@ -1,194 +1,195 @@
...
         entry,
         metadata,
         staticAccountIds,
-        btcAmount: debitAmount,
-        usdAmount: amount,
-      })
+        btcAmount: btcWithFee,
+        usdAmount: usdWithFee,
+      }
+      withdrawUsdFromDealer(dealerData)
```
Comment (dolcalmi) [inline src/services/ledger/domain/entry-builder.ts:172]: same here

File: src/services/ledger/domain/index.types.d.ts [hunk 6]
```diff
@@ -3,6 +3,11 @@ type LedgerAccountId = string & { [ledgerAccountId]: never }
 
 type TxMetadata = any
```
Comment (dolcalmi) [inline src/services/ledger/domain/index.types.d.ts:4]: as far as I remember we did not have this alert. could you please add an eslint exception?

File: src/services/ledger/domain/legacy-entry-builder.ts [hunk 7]
```diff
@@ -0,0 +1,247 @@
+import { AmountCalculator, WalletCurrency } from "@domain/shared"
```
Comment (dolcalmi) [inline src/services/ledger/domain/legacy-entry-builder.ts:1]: do we plan to remove this file? else, please rename

**Review by dolcalmi — DISMISSED** [review]

File: src/services/ledger/domain/legacy-entry-builder.ts (hunk 7 above)
Comment (UncleSamtoshi) [reply src/services/ledger/domain/legacy-entry-builder.ts:1]: Yes we plan on removing this file.

File: src/services/ledger/domain/entry-builder.ts (hunk 4 above)
Comment (UncleSamtoshi) [reply src/services/ledger/domain/entry-builder.ts]: @bodymindarts This is Juan's only hold up with the PR. I believe the concern is that when we are using a separate method (now named ```AddUsdToBtcConversionToEntry```)  to make changes to the entry, we are obscuring that the entry is being changed. @dolcalmi by returning the entry from ```AddUsdToBtcConversionToEntry``` and changing the name, does this address your problem?

Comment (bodymindarts) [reply src/services/ledger/domain/entry-builder.ts]: I think I understand @dolcalmi comment and think that the renaming is the easiest way to address it. I wouldn't use a capitalised name however (as this typically indicates a kind of constructor). `addUsdToBtcConvertionSoEntry` seems good however.

**Review by bodymindarts — APPROVED** [review]

**Pull request merged**